### ISSUE TITLE
No default backend found

### ISSUE THREAD

**Issue by rcsheong** [SublimeHaskell/SublimeHaskell#395]
Just installed SublimeHaskell and the Haskell Platform, ghc v8.2.2, cabal 2.0.0.1 (compiled using version 2.0.1.1 of the Cabal library), ST3 Build 3143, SublimeHaskell v2.1.9

I'm on Windows 10, hsdev is in %APPDATA%/cabal/bin

Whenever I start up ST3, I get this error message:
**No default backend found. Using the 'none' backend.**

I've tried editing my add_to_PATH to include "%APPDATA%\\cabal\\bin" and "%APPDATA%/cabal/bin" like so, with no effect:

    "add_to_PATH": [ 
            "%APPDATA%\\cabal\\bin",
            "%APPDATA%/cabal/bin"
       ],

I wondered if the "cabal global" builder was in a nonstandard location, and noticed the  "%PROGRAMFILES%/Haskell/bin" directory doesn't exist on my machine but "%PROGRAMFILES%/Haskell Platform/8.2.2/bin" does, so I tried including "C:\\Program Files\\Haskell Platform\\8.2.2\\bin" in the add_to_PATH as well but this made no difference.

**Comment by bscottm** [comment]
@rcsheong: _SublimeHaskell_ does look in `%APPDATA%\cabal\bin` as one of its well known places. I'm nowhere near a Windows machine at the moment, so it'll be a couple of days before I can diagnose. In the meantime, though, if you change the `log` _SublimeHaskell_ preference (in your user preferences) to `4`, you'll get a lot of diagnostic output that could help me help you.

Also, what happens when you execute `%APPDATA%\cabal\bin\hsdev version` from the `cmd` or PowerShell window? Even if _SublimeHaskell_ finds the executable, it tries to determine the `hsdev` version -- if there's an issue with running `hsdev`, that could also be the problem.

**Comment by rcsheong** [comment]
Changing the log to 4, how would I view the diagnostic output?

%APPDATA%\cabal\bin\hsdev version gives 0.3.0.1

I also got this error message right after installation, and it comes up sometimes if I have ST3 open for long enough, even if I haven't done anything with haskell in it
**No usable backends (hsdev, ghc-mod) found on PATH or
in the standard Haskell cabal and stack installation
locations.

Please check or update your SublimeHaskell user settings,
or install hsdev or ghc-mod.**

**Comment by bscottm** [comment]
@rcsheong: You can view the console and all of the debugging output by pressing &lt;Ctrl&gt;-\` (backtick.)

But worse, you have a version of `hsdev`that _SublimeHaskell_ does not yet support, officially. @mvoidex: You have an escaped version.

**Comment by bscottm** [comment]
@rcsheong: Try this:

    cabal install hsdev-0.2.5.1

That should install the current, usable version of `hsdev`.

**Comment by mylincode** [comment]
I ran into the same issue (on macOS), and assumed it was an issue with *SublimeHaskell* finding `hsdev`. It would be nice to have the most recent supported version in the README, and a more specific error message for version mismatches.

**Comment by bscottm** [comment]
@rcsheong: That's a good suggestion. I'll open a separate issue for that.

**Comment by MengPhantom** [comment]
@mylincode 
remove stack hsdev, and  `cabal install hsdev-0.2.5.1`
and this is my setting 
```

"backends": {
        "Local hsdev": {
            "default": true,
            "backend": "hsdev",
            "options": {
              "local": true,
              "port": 4567,
              "exec-with":"cabal",
              "install-dir":"/Users/a024/.cabal/bin"
            }
        },
        "No backend": {
            "default": false,
            "backend": "none"
        }
    }
```
you may want debug sublimehaskell with this user setting
```
{
	"log": 4
}

```
you can get debug message from console whit `ctrl ` `,and you can add logging msg with edit package file

```
Sublime Haskell: worker: SublimeHaskellStartBackend.do_startup
Sublime Haskell: Starting backend 'Local hsdev'
Sublime Haskell: HsDevBackend.__init__(localhost, 4567)
Sublime Haskell: ProcHelper.exec_with_wrapper: ['cabal', 'exec', 'hsdev', '--', 'version'] in /Users/a024/.cabal/bin
Sublime Haskell: normcmd :['/usr/local/bin/cabal', 'exec', 'hsdev', '--', 'version']
Sublime Haskell: popen_kwargs:{'stderr': -1, 'stdout': -1, 'cwd': '/Users/a024/.cabal/bin'}
Sublime Haskell: Starting local 'hsdev' server
Sublime Haskell: popen_kwargs:{'stderr': -1, 'stdout': -1, 'cwd': '/Users/a024/.cabal/bin'}
Sublime Haskell: hsdev initial: Server started at port 4567
Sublime Haskell: hsdev initial: 'hsdev' server started at port 4567
Sublime Haskell: Local 'hsdev' server started successfully.
Sublime Haskell: Connecting to 'hsdev' server at localhost:4567
Sublime Haskell: Host:localhost
Sublime Haskell: Port:4567
Sublime Haskell: Idx:0
Sublime Haskell: Retry:1
Sublime Haskell: Connections:<socket.socket object, fd=24, family=2, type=1, proto=6>
Sublime Haskell: Connect successful
```

**Comment by mylincode** [comment]
@MengPhantom You can also `stack install` a specific version, which is what I ended up doing.

Looking at the console is probably a good idea next time I run into this sort of thing. Though, I still feel that a better error pop-up would have been nice.

**Comment by bscottm** [comment]
@MengPhantom, @mylincode: Also, feel free to use environment variables and "~" for your home directory. I specifically tweaked the code to expand home directories and environment variables more than a year ago to do this. I got tired of absolute file names.

Also, you can use a `cabal sandbox` if `stack` isn't your preferred Haskell builder.

**Comment by bscottm** [comment]
2.1.23 should report a version number mismatch, if you inadvertently installed 0.3.x.y.

**Comment by ratijas** [comment]
### HOW TO install the correct version of hsdev in one command

`$ stack --resolver lts-9.1 --install-ghc install hsdev-0.2.5.1`

Guaranteed to work, since resolver `lts-9.1` was released the same day as `hsdev-0.2.5.1`.

(I think this one should be in the README or something)
